src/Microsoft.Data.SqlClient/tests/ManualTests/AlwaysEncrypted/ApiShould.cs

```diff
@@ -2114,6 +2117,93 @@ public void TestSqlCommandCancellationToken(string connection, int initalValue,
+        public void TestCustomKeyStoreProviderRegistration(string connectionString)
+        {
+            if (!SQLSetupStrategyAzureKeyVault.isAKVProviderRegistered)
+            {
+                SqlColumnEncryptionAzureKeyVaultProvider sqlColumnEncryptionAzureKeyVaultProvider =
+                    new SqlColumnEncryptionAzureKeyVaultProvider(new SqlClientCustomTokenCredential());
+
+                DummyKeyStoreProvider dummyProvider = new DummyKeyStoreProvider();
+
+                SqlConnection.RegisterColumnEncryptionKeyStoreProviders(customProviders:
+                    new Dictionary<string, SqlColumnEncryptionKeyStoreProvider>(capacity: 2, comparer: StringComparer.OrdinalIgnoreCase)
+                    {
+                        { "AZURE_KEY_VAULT", sqlColumnEncryptionAzureKeyVaultProvider},
+                        { "DummyProvider", dummyProvider}
+                    });
+
+                SQLSetupStrategyAzureKeyVault.isAKVProviderRegistered = true;
+            }
+
+            Dictionary<string, SqlColumnEncryptionKeyStoreProvider> requiredProvider =
+                new Dictionary<string, SqlColumnEncryptionKeyStoreProvider>()
+                {
+                    { "DummyProvider", new DummyKeyStoreProvider() }
+                };
+
+            Dictionary<string, SqlColumnEncryptionKeyStoreProvider> notRequiredProvider =
+                new Dictionary<string, SqlColumnEncryptionKeyStoreProvider>()
+                {
+                    { "DummyProvider2", new DummyKeyStoreProvider() }
+                };
+
+            string failedToDecryptErrorMessage = "Failed to decrypt a column encryption key using key store provider: " +
+                "'DummyProvider'. Verify the properties of the column encryption key and its column master key in your database.";
+            string providerNotImplementedMessage = "The method or operation is not implemented.";
+            string providerNotFoundMessage = "Invalid key store provider name: 'DummyProvider'. A key store " +
+                "provider name must denote either a system key store provider or a registered custom key store provider." +
+                " Valid system key store provider names are: 'MSSQL_CERTIFICATE_STORE', 'MSSQL_CNG_STORE', " +
+                "'MSSQL_CSP_PROVIDER'. Valid (currently registered) custom key store provider names are: 'DummyProvider2'.";
+
+            using (SqlConnection connection = new SqlConnection(connectionString))
+            {
+                connection.Open();
+
+                // will use DummyProvider in global cache
+                // provider will be found but it will throw when its methods are called
+                Exception ex = Assert.Throws<SqlException>(
+                      () => ExecuteQueryThatRequiresCustomKeyStoreProvider(connection));
+                Assert.Contains(failedToDecryptErrorMessage, ex.Message);
+                Assert.Contains(providerNotImplementedMessage, ex.Message);
```

Comment:
You can use `Equal` since this is the exact message you're expecting here—the same comment for other same assertions.

Reply:
The actual error message contains the last 10 bytes of the CEK, which is randomly generated. I could use regex but I think this is readable enough.

Reply:
Ok, wasn't aware we were aiming for culture independent test runnability since many other tests use hardcoded english messages. Updated this test.